### PR TITLE
Add Magic Leap Controller Support

### PR DESCRIPTION
**Pull request by atarng-magicleap**
This pull request addresses three issues.

One, defines the Magic Leap Component to support the new XR session profiles.
Two, addresses an issue where the controllers are being cached, despite closing the XRsession, and so when requesting a new session, chromium/webxr rejects the session due to the previous, now invalid, session id stored in the controller.
Three, based on available documentation for XRHandedness, there is support for input devices that do not specify a specific "handedness". This should be accounted for and allow these input events to still send button events.

References:
https://immersive-web.github.io/webxr-reference/webxr-device-api/xrinputsource.html
https://chromium.googlesource.com/chromium/src/+/ab6c1c84a115d5a05c6827de7a33b70fbbb8e5ea/device/vr/public/mojom/vr_service.mojom

### CODE REVIEW

File: src/components/magicleap-controls.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+ * 0 - trigger (intensity value from 0.5 to 1)
+ * 1 - grip
+ * 2 - touchpad
+ * 3 - menu (dispatch but better for menu options)
+ *
+ * Axis:
+ * 0 - touchpad x axis
+ * 1 - touchpad y axis
+ */
+var INPUT_MAPPING_WEBVR = {
```
Comment (dmarcos) [inline src/components/magicleap-controls.js]: Do you think we can remove WebVR references? My understanding is that there are no legacy WebVR browsers with magicleap support

Comment (atarng-magicleap) [reply src/components/magicleap-controls.js]: Yes, I missed this! Thanks for catching it!

File: src/systems/tracked-controls-webxr.js [hunk 2]
```diff
@@ -19,7 +19,16 @@ module.exports.System = registerSystem('tracked-controls-webxr', {
   updateControllerList: function () {
     var xrSession = this.el.xrSession;
     var self = this;
-    if (!xrSession) { return; }
+    if (!xrSession) {
```
Comment (dmarcos) [inline src/systems/tracked-controls-webxr.js]: Maybe better:

```
if (!xrSession) { 
 if (this..oldControllersLength === 0) { return; }
 this.oldControllersLength = 0;
 this.controllers = [];
 this.el.emit('controllersupdated', undefined, false); 
}

File: src/components/tracked-controls-webxr.js [hunk 3]
```diff
@@ -62,17 +62,26 @@ module.exports.Component = registerComponent('tracked-controls-webxr', {
   },
 
   emitButtonDownEvent: function (evt) {
-    if (!this.controller || evt.inputSource.handedness !== this.data.hand) { return; }
+    if (!this.controller) { return; }
```
Comment (dmarcos) [inline src/components/tracked-controls-webxr.js]: Why is this needed?

File: src/components/tracked-controls-webxr.js [hunk 4]
```diff
@@ -62,17 +62,26 @@ module.exports.Component = registerComponent('tracked-controls-webxr', {
...
+
     this.selectEventDetails.state.pressed = true;
     this.el.emit('buttondown', this.selectEventDetails);
     this.el.emit('buttonchanged', this.selectEventDetails);
     this.el.emit('triggerdown');
   },
 
   emitButtonUpEvent: function (evt) {
-    if (!this.controller || evt.inputSource.handedness !== this.data.hand) { return; }
+    if (!this.controller) { return; }
```
Comment (dmarcos) [inline src/components/tracked-controls-webxr.js]: Why is this needed? What's particular about magic leap with respect to other controllers?

Comment (dmarcos) [reply src/components/tracked-controls-webxr.js]: Is magicleap controller always returning `none` as handness or is it user configurable too like Oculus Go?

File: src/components/magicleap-controls.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  },
+
+  onModelLoaded: function (evt) {
+    var controllerObject3D = evt.detail.model;
+    // our glb scale is too large.
+    controllerObject3D.scale.set(0.01, 0.01, 0.01);
+  },
+
+  onAxisMoved: function (evt) {
+    emitIfAxesChanged(this, this.mapping[this.data.hand].axes, evt);
```
Comment (atarng-magicleap) [inline src/components/magicleap-controls.js]: no need for handedness here (in fact, this is undefined).

File: src/components/tracked-controls-webxr.js (hunk 4 above)
Comment (atarng-magicleap) [reply src/components/tracked-controls-webxr.js]: regarding this, we currently do not specify a handedness for our controllers (although, in theory it is possible for us to recognize which hand is currently holding the controller, and set it at runtime; however, we currently do not have that planned as a feature for our controllers).

Since Chromium Code enumerates a NONE state, and the webxr reference does mention a value of "" is possible, it is my personal believe that there should be handling for such a scenario where it is neither left nor right. (I'm actually not sure why the reference page lists an empty string, since it seems like it's going to set to none).

File: src/components/tracked-controls-webxr.js [hunk 6]
```diff
@@ -62,17 +62,24 @@ module.exports.Component = registerComponent('tracked-controls-webxr', {
   },
 
   emitButtonDownEvent: function (evt) {
-    if (!this.controller || evt.inputSource.handedness !== this.data.hand) { return; }
-    if (this.controller.gamepad) { return; }
+    if (!this.controller || this.controller.gamepad) { return; }
```
Comment (dmarcos) [inline src/components/tracked-controls-webxr.js]: Maybe factor out this code and the one below in to a method `isControllerPresent`?

**Comment by dmarcos** [comment]
In case you haven't seen it. I left a comment on https://github.com/aframevr/assets/pull/37 that might affect this PR.

**Comment by dmarcos** [comment]
Just in case it's not too hard. Do you have a short video you can share of Helio and A-Frame using hand-controls to tweet about once this lands?

**Comment by atarng-magicleap** [comment]
Of course! I'll go ahead prepare a video today!

File: src/components/tracked-controls-webxr.js (hunk 6 above)
Comment (dmarcos) [reply src/components/tracked-controls-webxr.js]: I think if we factor this code out the PR is ready to go. Thanks for the patience.

Comment (atarng-magicleap) [reply src/components/tracked-controls-webxr.js]: okay, yes I think I can do so!

**Comment by dmarcos** [comment]
Thank you! Great effort and thanks again for the patience!

**Comment by Utopiah** [comment]
@atarng-magicleap can you please share a link to the video example?
